src/gateway/Gateway.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC20Permit} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+
+/// @title ECG Gateway
+/// @notice Gateway to interract via multicall with the ECG
+/// Owner can select which calls are allowed
+abstract contract Gateway is Ownable, Pausable {
+    /// @notice Emitted when an external call fails with an error.
+    error CallExternalError(bytes innerError);
+
+    // never allow transferFrom to be whitelisted
+    // this avoid human error
+    bytes4 public immutable transferFromSelector =
+        bytes4(keccak256("transferFrom(address,address,uint256)"));
+
+    /// @notice emitted when a call is allowed or not by the function allowCall
+    event CallAllowed(
+        address indexed target,
+        bytes4 functionSelector,
+        bool isAllowed
+    );
+
+    /// @notice Stores calls to be executed after receiving a Balancer flash loan.
+    /// @dev The StoredCalls should only be set in the 'multicallWithBalancerFlashLoan' function
+    ///      Which is "onlyOwner" so the StoredCalls can only be set by the owner of the contract
+    bytes[] internal _storedCalls;
+
+    address internal _originalSender = address(1);
+
+    /// @notice mapping of allowed signatures per target address
+    /// For example allowing "flashLoan" on the balancer vault
+    mapping(address => mapping(bytes4 => bool)) public allowedCalls;
+
+    constructor() Ownable() Pausable() {}
+
+    /// @notice set pausable methods to paused
+    function pause() public onlyOwner {
+        _pause();
+    }
+
+    /// @notice set pausable methods to unpaused
+    function unpause() public onlyOwner {
+        _unpause();
+    }
+
+    /// @notice allow (or disallow) a function call for a target address
+    function allowCall(
+        address target,
+        bytes4 functionSelector,
+        bool allowed
+    ) public onlyOwner {
+        require(
+            functionSelector != transferFromSelector,
+            "Gateway: cannot allow transferFrom"
+        );
+
+        allowedCalls[target][functionSelector] = allowed;
+        emit CallAllowed(target, functionSelector, allowed);
+    }
+
+    /// @notice Executes an external call to a specified target.
+    ///         Only allows external calls to allowed target and function selector
+    ///         these whitelisted calls are stored in the factory
+    /// @dev    this function is only callable by the owner or if the owner initiated a flashloan (and then the flashloan contract called the contract back to continue the execution)
```

Comment:
I don't think this comment is up to date anymore

ps: even without `sweep()`, if there are tokens sitting on the contract, anyone can call callExternal(token, token.transfer(user, amount)) to sweep tokens

Reply:
Not if we don't allow transfer in the allowedCalls.

Reply:
btw yeah all the comments are outdated

but this time I waited to finish the dev before ahah